### PR TITLE
Bug fix/bts 1570

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

*This implements handling of Satellite Collections in Replication2.*

- [x] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [ ] **Regression tests**
  - [x] C++ **Unit tests** (See EE part)
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1364
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Replication2/Supervision/CollectionGroupSupervision.cpp [hunk 1]
```diff
@@ -197,30 +221,34 @@ auto createCollectionGroupTarget(
...
+
+      ADB_PROD_ASSERT(group.targetCollections.contains(cid))
+          << "collection " << cid << " is listed in collection group "
+          << group.target.id << " but not in Target/Collections";
+      for (size_t k = 0; k < shardList.size(); ++k) {
+        ResponsibleServerList serverids{};
+        auto const& log = replicatedLogs[spec.shardSheaves[k].replicatedLog];
+        serverids.servers.emplace_back(log.leader.value());
+        for (auto const& p : log.participants) {
+          if (p.first != log.leader) {
```
Comment (apetenchea) [inline arangod/Replication2/Supervision/CollectionGroupSupervision.cpp:242]: A note for myself.: if the log's leader in Target is failed, the supervision will elect a new one, but this one remains set in Target.

**Review by apetenchea — APPROVED** [review]

File: arangod/Replication2/Supervision/CollectionGroupSupervision.cpp [hunk 2]
```diff
@@ -273,8 +301,13 @@ auto pickBestServerToRemoveFromLog(
                      // then remove leaders that are not Target leaders
                      return compareTuple(left) < compareTuple(right);
                    });
+#if false
+  // TODO i think this assertion is not correct. It is possible that the leader
```
Comment (apetenchea) [inline arangod/Replication2/Supervision/CollectionGroupSupervision.cpp:305]: I agree. Removing the target leader should be fine. Even if the removed server remains set in target, the supervision will handle this in `checkLeaderSetInTarget`.

**Review by apetenchea — APPROVED** [review]

File: arangod/Replication2/Supervision/CollectionGroupSupervision.cpp [hunk 3]
```diff
@@ -69,14 +69,38 @@ auto computeEvenDistributionForServers(
...
+    std::vector<ParticipantId> allParticipants;
+    allParticipants.reserve(health._health.size());
+    for (auto const& [p, h] : health._health) {
+      allParticipants.push_back(p);
+    }
+    {
+      // TODO reuse random device?
+      std::random_device rd;
+      std::mt19937 g(rd());
+      std::shuffle(allParticipants.begin(), allParticipants.end(), g);
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Replication2/Supervision/CollectionGroupSupervision.cpp:84]: Good spot!!
Removed it in both places.

File: arangod/Replication2/Supervision/CollectionGroupSupervision.cpp [hunk 4]
```diff
@@ -197,30 +221,34 @@ auto createCollectionGroupTarget(
...
+    auto const& targetCollection = group.targetCollections.at(cid);
+
+    // If we have shadow collections we do not have any shards
+    if (!targetCollection.immutableProperties.shadowCollections.has_value()) {
+      std::generate_n(std::back_inserter(shardList),
+                      attributes.immutableAttributes.numberOfShards, [&] {
+                        return basics::StringUtils::concatT("s", uniqid.next());
+                      });
+
+      ADB_PROD_ASSERT(group.targetCollections.contains(cid))
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Replication2/Supervision/CollectionGroupSupervision.cpp:234]: True

**Pull request merged**